sdk/identity/azure-identity/src/token_cache.cpp

```diff
@@ -0,0 +1,148 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+#include "private/token_cache.hpp"
+#include "private/token_cache_internals.hpp"
+
+#include <chrono>
+#include <mutex>
+
+using Azure::Identity::_detail::TokenCache;
+
+using Azure::Core::Credentials::AccessToken;
+
+decltype(TokenCache::Internals::Cache) TokenCache::Internals::Cache;
+decltype(TokenCache::Internals::CacheMutex) TokenCache::Internals::CacheMutex;
+
+#if defined(TESTING_BUILD)
+std::function<void()> TokenCache::Internals::OnBeforeCacheWriteLock;
+std::function<void()> TokenCache::Internals::OnBeforeItemWriteLock;
+#endif
+
+namespace {
+bool IsFresh(
+    std::shared_ptr<TokenCache::Internals::CacheValue> const& item,
+    std::chrono::system_clock::time_point now)
+{
+  // If cached token expires in less than MinExpiry from now, it's cached value won't be returned,
+  // newer value will be requested.
+  constexpr auto MinExpiry = std::chrono::minutes(3);
+
+  return item->AccessToken.ExpiresOn > now + MinExpiry;
+}
+
+std::shared_ptr<TokenCache::Internals::CacheValue> GetOrCreateValue(
+    TokenCache::Internals::CacheKey const& key)
+{
+  {
+    std::shared_lock<std::shared_timed_mutex> cacheReadLock(TokenCache::Internals::CacheMutex);
+
+    auto const found = TokenCache::Internals::Cache.find(key);
+    if (found != TokenCache::Internals::Cache.end())
+    {
+      return found->second;
+    }
+  }
+
+#if defined(TESTING_BUILD)
+  if (TokenCache::Internals::OnBeforeCacheWriteLock != nullptr)
+  {
+    TokenCache::Internals::OnBeforeCacheWriteLock();
+  }
+#endif
+
+  std::unique_lock<std::shared_timed_mutex> cacheWriteLock(TokenCache::Internals::CacheMutex);
+
+  // Search cache for the second time, in case the item was inserted between releasing the read lock
+  // and acquiring the write lock.
+  auto const found = TokenCache::Internals::Cache.find(key);
+  if (found != TokenCache::Internals::Cache.end())
+  {
+    return found->second;
+  }
+
+  // Clean up cache from expired items (once every N insertions).
+  {
+    auto const cacheSize = TokenCache::Internals::Cache.size();
+
+    // N: cacheSize (before insertion) is >= 32 and is a power of two.
+    if (cacheSize >= 32 && (cacheSize & (cacheSize - 1)) == 0)
```

Comment:
Where did 32 come from?

You might want to explain what X&(X-1) does, many people don't understand that particular trick (especially since it only works on processors which implement 2s complement arithmetic).

Reply:
32 is just a power of two I chose instinctively. It can be any number.
As of the "power of two" explanation, the comment just above `if()` says that:
`"N: cacheSize (before insertion) is >= 32 and is a power of two."`

Reply:
My 2nd comment is actually a *about* the "and is a power of 2" comment. *I* know that X&(X-1) reduces the number of bits in a integer value by 1 (on 2s complement CPUs). But I don't believe that is general knowledge. My comment is that you might want to add a link to something which explains the trick, because to me, it was *NOT* obvious (I once spent *hours* trying to figure out why it worked).

I know it because it was one of the solutions to the "Count the number of 1 bits in a 32bit integer" problem, but otherwise that would be totally mystifying code.

Reply:
I'll add a link. He best book for such tricks that I know is "Hacker's Delight" by Hentry S. Warren, but it is not freely available online. I'll find a good online article.
BTW, starting with C++20, two's compliment is required.